opennft/opennft.py

```diff
@@ -468,6 +505,32 @@ def onChangeDataType(self):
         self.cbgetMAT.setChecked(self.cbgetMAT.isChecked() and self.cbDataType.currentText() == 'DICOM')
         self.cbgetMAT.setEnabled(self.cbDataType.currentText() == 'DICOM')
 
+    # --------------------------------------------------------------------------
+    def showPluginDlg(self):
+        if self.pluginWindow.exec_():
+            self.plugins = []
+            for p in range(len(self.pluginWindow.plugins)):
+                if self.pluginWindow.lvPlugins.model().item(p).checkState():
+                    self.plugins += [{'module': self.pluginWindow.plugins[p]}]
+
+    # --------------------------------------------------------------------------
+    def initializePlugins(self):
+        for i in range(len(self.plugins)):
+            if len(self.plugins[i]['module'].META['plugin_init']):
+                self.plugins[i]['object'] = eval("self.plugins[i]['module']." + self.plugins[i]['module'].META['plugin_init'].format(**self.P))
+    
+    # --------------------------------------------------------------------------
+    def finalizePlugins(self):
+        for i in range(len(self.plugins)):
+            self.plugins[i]['object'] = None
+    
+    # --------------------------------------------------------------------------
+    def updatePlugins(self):
+        for i in range(len(self.plugins)):
+            m = self.plugins[i]['module'].META
+            if (self.recorder.getLastEvent() == eval("erd.Times." + m['plugin_time'])) and eval(m['plugin_signal']):
+                exec("self.plugins[i]['object']." + m['plugin_exec'])
+    
```

Comment:
I have tested new version, but faced a new bug.
1. Opened Plugin window, chose plugins and clicked OK
2. Pressed Setup
3. Opened Plugin window one more time (for example, to be sure I chose plugins)
4. Pressed Start after that and caught an error

Traceback (most recent call last):
  File "C:\OpenNFT\opennft\opennft.py", line 815, in callMainLoopIteration
    self.updatePlugins()
  File "C:\OpenNFT\opennft\opennft.py", line 532, in updatePlugins
    exec("self.plugins[i]['object']." + m['plugin_exec'])
  File "<string>", line 1, in <module>
KeyError: 'object'

So, I see two ways of resolving thise problem.
- disable Plugin button after Setup (don't think it's a good idea)
- initialize plugins when user pressed Ok and do nothing when Cancel

Reply:
How about a third way?
- disable Start if clicking on Plugin

As far as I can see, there is a reason why we have had these buttons (Initialize - Review parameters - Setup - Start) in this order, and I put Plugin before Setup for that same reason. Therefore, if one clicks on Plugin again, the (s)he should click on Setup first before going to Start. 
Besides, plugins usually relies on P subfields updated during Setup. See last line:
https://github.com/OpenNFT/OpenNFT/blob/95986523518dd94f01a1f94335bf8ca1192e65d9/opennft/opennft.py#L517-L520. Therefore, plugins must be initialized during Setup.

Moreover, initializePlugins can be also made more eloquent on which plugins have been chosen and initialized, so you do not have to go back to Plugin

Reply:
I have now implemented the third way. Moreover, I have disabled the Plugin button at the beginning, and it will be enabled only after Initialization.

Reply:
Checked. Seems to be working fine.

What do you think of disabling Plugin button during the run? User can open window during the run and ruin it. 

Or may be it's too much and I became too prudent.

Reply:
Good idea! I think you are right. I also prefer fool-proofing. :)

Reply:
We should probably do the same with the Setup button.